pool/native.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package pool
+
+import (
+	"reflect"
+	"unsafe"
+)
+
+// NativePoolOptions specify options for NativePool.
+type NativePoolOptions struct {
+	Construct func(ptr interface{})
+	Size      uint
+	Type      reflect.Type
+}
+
+// NativePool represents an object pool which is opaque to the Go GC.
+type NativePool interface {
+	Get() interface{}
+	GetOr(OverflowFn) interface{}
+	Put(interface{})
+
+	// Owns determines if the object belongs to the pool.
+	Owns(interface{}) bool
+
+	// Size returns the used and the total capacity of the pool.
+	Size() (uint, uint)
+}
+
+// OverflowFn produces non-pooled objects.
+type OverflowFn func() interface{}
+
+// NewNativePool constructs a new NativePool.
+func NewNativePool(opts NativePoolOptions) NativePool {
+	p := &nativePool{
+		free: make(chan uint, opts.Size),
+		opts: opts,
+		// TODO(@kobolog): alignment & padding.
+		step: uint(hsz + opts.Type.Size())}
+
+	p.size = p.opts.Size * p.step
+
+	if p.size == 0 {
+		panic("native-pool: pool size is zero")
+	}
+
+	if p.step%uint(p.opts.Type.Align()) != 0 {
```

Comment:
So the compiler pads types to be properly aligned, which means both `hdr` and whatever type was passed in `opts` both are supposed to be fine automatically. So I guess this invariant could be only broken if those alignments are somehow incompatible, e.g. `alignof(hdr)` < `alignof(T)`. I don't know whether this could happen or not, so I added this check just in case.

Actually, I'm gonna change `hdr.idx` to uint64 so that it has the highest alignment, which probably should always work.